numba/unicode.py

```diff
@@ -745,6 +746,42 @@ def concat_impl(a, b):
         return concat_impl
 
 
+@njit
+def _repeat_impl(str_arg, mult_arg):
+        if str_arg == '' or mult_arg < 1:
+            return ''
+        elif mult_arg == 1:
+            return str_arg
+        else:
+            new_length = str_arg._length * mult_arg
+            new_kind = str_arg._kind
+            result = _empty_string(new_kind, new_length)
+            # make initial copy into result
+            len_a = len(str_arg)
+            _strncpy(result, 0, str_arg, 0, len_a)
+            # loop through powers of 2 for efficient copying
+            floor_log = floor(log(mult_arg) / log(2))
```

Comment:
Was wondering if this can be done using shifts instead of relying on floating point working nicely... how does this look? Suggestions/improvements welcomed:
```python
            # make initial copy into result
            len_a = len(str_arg)
            _strncpy(result, 0, str_arg, 0, len_a)
            # loop through powers of 2 for efficient copying
            jmp = 1
            amount = jmp * len_a
            while True:
                newjmp = jmp << 1
                if (newjmp * len_a) <= new_length:
                    _strncpy(result, amount, result, 0, amount)
                    jmp = newjmp
                    amount = jmp * len_a
                else:
                    break
            else:
                # exact multiple
                return result

            # complete the rest of the copies
            write_ptr = amount
            _strncpy(result, write_ptr, result, 0, (new_length - write_ptr))
            return result
```

Reply:
do you mind if I avoid using `<<`? What do you think about this? I think it implements more directly the idea of repeatedly doubling the size of the copy area. Plus it's shorter and avoids the while...else.

```python
            copy_size = len_a
            while 2 * copy_size <= new_length:
                # if copying won't take me outside the new length
                _strncpy(result, copy_size, result, 0, copy_size)
                copy_size *= 2

            if not 2 * copy_size == new_length:
                # if copy_size not an exact multiple it then needs
                # to complete the rest of the copies
                rest = new_length - copy_size
                _strncpy(result, copy_size, result, copy_size - rest, rest)
                return result
```

Reply:
oh not at all, was just wanting to avoid use of floating point for loop bounds, if you've got something simpler/equivalent but different that works then that's fine